### ISSUE TITLE
hash literals always procude hash-equal?

### ISSUE THREAD

**Issue by cderici** [pycket/pycket#237]
$ pycket --new
\> #hash()
#hash()
\> #hasheq() 
#hash()
\> #hasheqv()
#hash()

This causes issues like `for/hasheqv` to procude `hash-equal?` instead of an `hash-eqv?`, (which in turn generates a different serialization for Pycket than Racket).

The linklet coming to the `compile-linklet` is:

'(linklet ((.top-level-bind! .top-level-require!) (.mpi-vector .syntax-literals) (.namespace .phase .self .inspector .bulk-binding-registry .set-transformer!)) () (quote #hash()))'

which tells me that we're doing something wrong that causes `read` to fail (to procude the correct one).

I confirmed that we're using the expander's `read` instead of the rpython one.

**Comment by cderici** [comment]
when `#hasheqv()` is called, for instance, the constructor `make-immutable-hasheqv` (called within `read-hash` in the expander) produces a correct value (which is `W_EqvImmutableHashTable`), but somewhere after that it changes to `W_EqualHashTable`, as I observed that the `current-print` in the REPL gets a `W_EqualHashTable` as an input (and does print it correctly).

**Comment by cderici** [comment]
Ok I think I know what's going on. The real issue is neither of Pycket's expanders (`zo-expand.rkt`, nor `expand.py`) care about any hash other than `hash-equal?`. For example, evaluate the following module on both new and old pyckets 

`(module hasheqv '#%kernel
(printf "h : ~a -- (hash-eqv? h): ~a\n" #hasheqv() (hash-eqv? #hasheqv())))`

and you'll get `h : #hash() -- (hash-eqv? h): #f`, which is clearly wrong.

Here's how I got here: I tracked the error in the previous comment to a `syntax->datum` call (in the Racket expander) that's happening when the read value is wrapped right after `read-hash`, and that calls `syntax-map` -> `datum-map` -> `datum-map-slow`, which uses `for/hasheqv` to actually create the hash value, which fully expands to the code that uses `#hasheqv`, which in turn encoded in the expander linklet with `hash-keys/hash-vals` AST, which totally loses the `eq/eqv` information (`expand.py` always produces `W_EqualHashTable`).

I think extending the hash-table json AST with the eq info and adding that to both `zo-expand.rkt` and `expand.py`  will solve this issue.
